nnsvg.c

```diff
@@ -0,0 +1,169 @@
+
+    // Get an instantiated NSVGimage struct from the parsed input
+    // and store its address in our userdata
+    if ( is_svg_data ) {
+        *udata = nsvgParse((char*)input, "px", dpi);
+    }
+    else {
+        *udata = nsvgParseFromFile(input, "px", dpi);
+    }
+    if ( !*udata ) {
+        lua_pushstring(L, "Failed parsing SVG.");
+        lua_error(L);
+    }
+    return 1; // Return this new userdata
+}
+
+NSVGimage * check_NSVGimage(lua_State * L, int n) {
+    // This checks that the thing at n on the stack is a correct nnSVGImage 
+    // wrapping userdata (tagged with the "luaL_nnSVGImage" metatable).
+    NSVGimage * image = *(NSVGimage **)luaL_checkudata(L, n, NNSVG_METATABLE_NAME);
+    return image;
+}
+
+// ==============================================
+// Lua wrapping methods
+
+// Exported as :free() (to be called explicitely if we want to free
+// the NSVGimage early) and as :__gc() (called by the async Lua GC).
+static int nnSVGImage_free(lua_State *L) {
+    NSVGimage ** udata = (NSVGimage **)luaL_checkudata(L, 1, NNSVG_METATABLE_NAME);
+    // printf("nnSVGImage_free %p\n", *udata);
+    if ( *udata )
+        nsvgDelete(*udata);
+    *udata = NULL;
+    return 0;
+}
+
+// Get native width and height of SVG image
+static int nnSVGImage_getSize(lua_State *L) {
+    NSVGimage * image = check_NSVGimage(L, 1);
+    lua_pushinteger(L, image->width);
+    lua_pushinteger(L, image->height);
+    return 2;
+}
+
+// Rasterize SVG image on the provided BlitBuffer (which should
+// be already sized and allocated)
+static int nnSVGImage_drawTo(lua_State *L) {
+    NSVGimage * image = check_NSVGimage(L, 1);
+
+    // 2nd argument should be a BlitBuffer instance
+    // We expect it to be a luajit ffi cdata, but the C API does not have a #define for
+    // that type. But it looks like its value is higher than the greatest LUA_T* type.
+    if ( lua_type(L, 2) <= LUA_TTHREAD ) {// Higher plain Lua datatype (lua.h)
+	luaL_typerror(L, -1, "BlitBuffer");
+    }
+    BlitBuffer * bb = (BlitBuffer*) lua_topointer(L, 2);
+
+    // nanosvg's rasterizer won't automatically scale the image
+    // to adjust it to the target buffer w/h.
+    // But it allows a scale factor that is applied to both axis
+    // (so, it forces us to keep the original aspect ratio, we
+    // can't have it stretch SVGs).
+    // The positionning of the original (possibly scaled) SVG image
+    // inside the already sized and allocated BlitBuffer has to be
+    // done by caller, which may provide these optional arguments:
+    const float scale_factor = (float)luaL_optnumber(L, 3, 1.0);
+    const int offset_x = (int)luaL_optint(L, 4, 0);
+    const int offset_y = (int)luaL_optint(L, 5, 0);
+
+    NSVGrasterizer *rast = nsvgCreateRasterizer();
+    if (rast == NULL) {
+        lua_pushstring(L, "Could not init rasterizer.");
+        lua_error(L);
+    }
+    nsvgRasterize(rast, image, offset_x, offset_y, scale_factor, bb->data, bb->w, bb->h, 4*bb->w);
```

Comment:
The final argument ought to be `bb->stride`, with an earlier check to make sure sure that the dest BB type is BBRGB32 ;).

Reply:
Is there a proper way to check for BBRGB32 ?
`#define TYPE_BBRGB32 5` is in blitbuffer.c, not in blitbuffer.h.
Should I just check for `5`, or move the #define in blitbuffer.h (and which, only the TYPE_BB*?)

Reply:
Oh, but there's no type in the structs:) So, what is the proper way to be sure it's BBRGB32?

Reply:
```c
#define MASK_TYPE 0xF0
#define SHIFT_TYPE 4
#define TYPE_BB4 0
#define TYPE_BB8 1
#define TYPE_BB8A 2
#define TYPE_BBRGB16 3
#define TYPE_BBRGB24 4
#define TYPE_BBRGB32 5
#define GET_BB_TYPE(bb) (((MASK_TYPE & bb->config) >> SHIFT_TYPE))
```

Reply:
You're including `"blitbuffer.h"` already, right?

You should be able to do `if (GET_BB_TYPE(bb) == TYPE_BBRGB32)` ;).

Reply:
Oh, yeah, right, it's in the C file, not the header.

Well, that's a mistake :D.

Reply:
So, just use stride for now, I'll take care of the rest later ;).